doc/compiler/optimizer/BenefitInliner.md

```diff
@@ -0,0 +1,94 @@
+3. Commit the inlining plan and inline the methods.
+    
+## Higher-Level Structure of the parts of BenfitInliner
+
+### 1. BenefitInliner.cpp
+
+Class hierarchy of the adapter class:
+* **TR::Optimization**
+    * **BenefitInlinerWrapper**
+
+Class hierarchy of the real inliner:
+* **TR_InlinerBase**
+    * **BenefitInlinerBase**
+       * **BenefitInliner**
+
+* **BenefitInlinerWrapper** is an adapter class which extends from **TR::Optimization**. `BenefitInlinerWrapper::perform()` initializes a **BenefitInliner** and controls how benefit inlining process is being done.
+
+* **BenefitInliner** is the real Inliner that contains functionalities for selecting inlining candidates. When a **BenefitInliner** is initialized, it comes with an inlining budget that constrains inlining by avoiding an uncontrollable increase in binary size. The inlining budget will be calculated based on the hotness and size of the method being JIT compiled. `BenefitInliner::buildingInliningDependecyTree()` builds an **IDT** that contains all the candidate methods to be inlined. `BenefitInliner::inlinerPacking()` consumes the **IDT** and selects the optimal set of methods to be inlined given the inlining budget and the cost and benefit of each method in the **IDT**. 
+
+* **BenefitInlinerBase** contains the functionalities of doing the actual inlining according to the inlining proposal proposed by `BenefitInliner::inlinerPacking()`. 
+
+### 2. InliningProposal.cpp
+
+* **InliningProposal** proposes an optimal set of methods that will be inlined. **BenefitInlinerBase** takes **InliningProposal** as a dependency. **InliningProposal** will be set by the `BenefitInliner::inlinerPacking()` and will be consulted during the inlining. 
+
+* **InliningProposalTable** is a 2D table of **InliningProposal**. By the nature of `BenefitInliner::inlinerPacking()` being a nested knapsack algorithm, a 2D table is required. 
+
+### 3. IDTNode.cpp, IDT.cpp & IDTBuilder.cpp
+
+* **IDTNode** is the building block of **IDT**. It is the abstract representation of a candidate method. 
+
+An **IDTNode** has the following important properties:
+> 1. **call ratio** - It estimates how many times this node is visited per the execution of its parent. 
+> 2. **root call ratio** - It estimates how many times this node is visited per the execution of the root node. 
+> 3. **Inlining method summary** - It captures potential optimization opportunities for inlining this method. (Refer to **InliningMethodSummary.cpp** for details)
+> 4. **static benefit** - The indirect benefit (optimizations unlocked) of inlining this method.
+> 5. **budget** - The remaining budget for the descendants of this node.
+> 6. **Bytecode Index** - The call site index of this method.
+
+The **benefit** of inlining a candidate method is calculated as: `Benefit = root call ratio * static benefit`
+
+The **cost** of inlining a candidate method is equal to the size of this method.
```

Comment:
Is this a strict requirement or can a language implementation choose different equations?

Reply:
This is proposed by the research paper